osquery/tables/system/linux/deb_packages.cpp

```diff
@@ -0,0 +1,185 @@
+// Copyright 2004-present Facebook. All Rights Reserved.
+
+#include <boost/algorithm/string.hpp>
+#include <osquery/tables.h>
+
+// see README.api of libdpkg-dev
+#define LIBDPKG_VOLATILE_API
+
+extern "C" {
+#include <dpkg/dpkg-db.h>
+// copy pasted from dpkg-db.h
+// these enums are inside struct pkginfo and are not visible for other headers
+enum pkgwant {
+  want_unknown,
+  want_install,
+  want_hold,
+  want_deinstall,
+  want_purge,
+  /** Not allowed except as special sentinel value in some places. */
+  want_sentinel,
+} want;
+/** The error flag bitmask. */
+enum pkgeflag {
+  eflag_ok = 0,
+  eflag_reinstreq = 1,
+} eflag;
+enum pkgstatus {
+  stat_notinstalled,
+  stat_configfiles,
+  stat_halfinstalled,
+  stat_unpacked,
+  stat_halfconfigured,
+  stat_triggersawaited,
+  stat_triggerspending,
+  stat_installed
+} status;
+
+#include <dpkg/dpkg.h>
+#include <dpkg/pkg-array.h>
+#include <dpkg/parsedump.h>
+}
+
+namespace osquery {
+namespace tables {
+
+// dpkg versions in 12.04 and 14.04 use completely different sorters
+// bringing this one forward to smoother over the incompatibility
+int pkg_sorter(const void *a, const void *b) {
+  const struct pkginfo *pa = *(const struct pkginfo **)a;
+  const struct pkginfo *pb = *(const struct pkginfo **)b;
+  const char *arch_a = pa->installed.arch->name;
+  const char *arch_b = pb->installed.arch->name;
+  int res;
+
+  res = strcmp(pa->set->name, pb->set->name);
+  if (res)
+    return res;
+
+  if (pa->installed.arch == pb->installed.arch)
+    return 0;
+
+  return strcmp(arch_a, arch_b);
+}
+
+void dpkg_setup(char const *name,
```

Comment:
Do we need to pass a `const *` if this isn't used anywhere but as an aux internal setup helper? Can we call `dpkg_set_progname` directly `dpkg_set_progname("osquery");`? And nit about the potentially confusing `array` variable name. ;)